java/dataset/src/main/java/org/apache/arrow/dataset/file/FileSystemDatasetFactory.java

```diff
@@ -26,14 +28,18 @@
  */
 public class FileSystemDatasetFactory extends NativeDatasetFactory {
 
+  private static final AtomicBoolean addedS3ShutdownHook = new AtomicBoolean(false);
```

Comment:
I added an atomic check to ensure we only register a single hook. Without it, we could register multiple hooks to call `EnsureS3Finalized`.

Reply:
Can we just stick it in JniLoader? https://github.com/apache/arrow/blob/main/java/dataset/src/main/java/org/apache/arrow/dataset/jni/JniLoader.java

Reply:
Ah yes, this is a much better spot for this. Will update. The atomic bool did feel like overkill